### ISSUE TITLE
finally() and socket.ref() undefined

### ISSUE THREAD

**Issue by sdetweil** [serviejs/popsicle-transport-http#7]
I am trying to upgrade the dependencies on a node/browser app  to current levels to get rid of a grpc problem. 

one, google-assistant, uses a module that uses popsicle 
```
└─┬ client-oauth2@4.3.3
  └─┬ popsicle@12.1.0
    └── popsicle-transport-http@1.1.2 
```

i call a client-oauth method to refresh my oauth2 token and now get an error that finally is not defined.. 

if i change the finally()'s to .then() (workaround)
i now get a : socket.ref is not a function
both from popsicle-transport-http

running node 14.15.3
in electron 11.3.0

i read in my old refresh token and then update it
```
				GA_Auth = new ClientOAuth2({
					clientId: keyinfo.installed.client_id,
					clientSecret: keyinfo.installed.client_secret,
					accessTokenUri: keyinfo.installed.token_uri,
					authorizationUri: keyinfo.installed.auth_uri,
					redirectUri: "http://localhost:5000/assistant_auth", 
					scopes: ["https://www.googleapis.com/auth/assistant-sdk-prototype"],
				});
...
					let tokeninfo = JSON.parse(
						fs.readFileSync(_gp.resolve(_ourpath, "token.json"))
					);
					var token = GA_Auth.createToken(
						tokeninfo.access_token,
						tokeninfo.refresh_token ? tokeninfo.refresh_token : null
					);

					// Set the token TTL.
					token.expiresIn(60 * 24 * 60); //seconds
					console.log("assistant refreshing access token on startup");
					// Refresh the users credentials and save the new access token and info.
					token
						.refresh()
						.then((user) => {
							console.log("assistant saving tokens on startup");
							saveTokens(user);
							AssistantInit(gaconfig, geoposition);
						})
						.catch((error) => {
							console.log("error on startup refresh ="+error);   //< code goes here on finally/socket.ref
						});
```

**Comment by blakeembrey** [comment]
Can you share the related stack track and full error messages? I can't really use the snippet you've included as it's using a bunch of things I cannot set up myself.

**Comment by sdetweil** [comment]
this is in the electron developers console... I don't get a typical stack trace 

```
error on startup refresh =TypeError: socket.ref is not a function  -- my then().catch()

controller.js:444 console.trace
(anonymous) @ controller.js:444
R @ sdk-3.0.0.js:8
N @ sdk-3.0.0.js:8
x @ sdk-3.0.0.js:8
O @ sdk-3.0.0.js:8
m @ sdk-3.0.0.js:8
characterData (async)
(anonymous) @ sdk-3.0.0.js:8
J @ sdk-3.0.0.js:8
D @ sdk-3.0.0.js:8
(anonymous) @ sdk-3.0.0.js:8
Promise.then (async)
T @ sdk-3.0.0.js:8
(anonymous) @ sdk-3.0.0.js:8
m @ sdk-3.0.0.js:8
characterData (async)
(anonymous) @ sdk-3.0.0.js:8
J @ sdk-3.0.0.js:8
then @ sdk-3.0.0.js:8
Request @ node.ts:210
fetch @ common.ts:20
request @ /home/sam/sm_new_updated/plugins/sm-Assistant/node_modules/client-oauth2/src/request/index.js:13
ClientOAuth2._request @ /home/sam/sm_new_updated/plugins/sm-Assistant/node_modules/client-oauth2/src/client-oauth2.js:273
ClientOAuth2Token.refresh @ /home/sam/sm_new_updated/plugins/sm-Assistant/node_modules/client-oauth2/src/client-oauth2.js:376
(anonymous) @ controller.js:436  
```
my code package,json
```
{
  "name": "sm-Assistant",
  "version": "1.0.0",
  "description": "google assistant for smart-mirror",
  "main": "controller.js",
  "scripts": {
    "postinstall": "./post_install",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "sam detweiler",
  "repository": "https://github.com/sdetweil/sm-Assistant",
  "license": "ISC",
  "dependencies": {
    "@grpc/grpc-js": "^1.2.11",
    "client-oauth2": "^4.3.0",
    "google-assistant": "git+https://git@github.com/sdetweil/google-assistant.git",
    "js-beautify": "^1.13.0",
    "prettier": "^2.0.5"
  },
  "devDependencies": {
    "electron-rebuild": "^1.11.0"
  }
}
```
this is the module tree
```
sm-Assistant@1.0.0 /home/sam/sm_new_updated/plugins/sm-Assistant
└─┬ client-oauth2@4.3.3
  └─┬ popsicle@12.1.0
    └── popsicle-transport-http@1.1.2 
```

the ref() error is because SocketSet class doesn't have this method.. 
so there must be some assumption about the runtime binding defaults.

but finally() must also be some ts to js compiler level dependency

**Comment by sdetweil** [comment]
this is not a react app... I see ref() mentioned relative to react.

**Comment by sdetweil** [comment]
also socket.once
line 576
```
                // Existing socket may already have negotiated ALPN protocol.
                if (socket.alpnProtocol != null)
                    return onConnect();
                socket.once("secureConnect", onConnect);
                socket.once("error", onError);
                socket.once("close", onClose);
```

**Comment by sdetweil** [comment]
this commit , Fix http2 connection on node.js ^12.17, introduced socket.once
this commit , Rewrite connection management for async sockets, introduced finally() and socket.ref()/socket.unref()

**Comment by blakeembrey** [comment]
Are you running this in the electron renderer or something? I tried testing this myself in electron and it works fine with https://github.com/facebook-atom/jest-electron-runner.

Additionally, the commits you refer to moved where `.ref()` is happening but it's always been there.

**Comment by blakeembrey** [comment]
There's also been zero changes around `.once`, "Fix http2 connection on node.js" didn't change anything around `.ref()` either. 

> the ref() error is because SocketSet class

How can you tell it's a `SocketSet` class? Node.js has had `.ref()` since v0.9.1, so there must be something else happening here if you have that class. It doesn't sound like a node.js class.

**Comment by blakeembrey** [comment]
Did you try installing an old version of `popsicle-transport-http` to verify this is a new issue? You can do `npm install popsicle-transport-http@1.0.9` to get the release before the commits you're referring to. Even in that version was a reference to `.ref()` though, here: https://github.com/serviejs/popsicle-transport-http/blob/000c3ec1d5fa29b7111943cb3e718f98c46df323/src/index.ts#L834. Which happened every time we made a HTTP request: https://github.com/serviejs/popsicle-transport-http/blob/000c3ec1d5fa29b7111943cb3e718f98c46df323/src/index.ts#L275. This would help me understand if it's a recent addition, although the APIs I use to create sockets have always been the same.

**Comment by blakeembrey** [comment]
Also, if you don't have `.once` on the socket this definitely isn't a node.js socket. It's coming from somewhere else. Are you bundling or polyfilling native node.js dependencies somehow?

**Comment by blakeembrey** [comment]
I realized `SocketSet` must be from within the package, but I've been unable to reproduce the issue locally. The `SocketSet` is a reference to the socket pool, but there's complete code coverage for the behaviors and there aren't any type issues where I've used the `SocketSet` as a `Socket`. I'll keep investigating tomorrow, but if you have a small reproduction I can use that'd be extremely helpful.

**Comment by sdetweil** [comment]
i installed 1.0.9 and no longer get any errors,  and the application works. 
there are special methods in 1.0.,9 for ref and unref, approx line 833 of the generated index.js
```
/**
 * Track socket usage.
 */
const SOCKET_REFS = new WeakMap<Socket | TLSSocket, number>();

/**
 * Track socket refs.
 */
function ref(socket: Socket | TLSSocket) {
  const count = SOCKET_REFS.get(socket) || 0;
  if (count === 0) socket.ref();
  SOCKET_REFS.set(socket, count + 1);
}

/**
 * Track socket unrefs and globally unref.
 */
function unref(socket: Socket | TLSSocket) {
  const count = SOCKET_REFS.get(socket);
  if (count) {
    if (count === 1) {
      socket.unref();
      SOCKET_REFS.delete(socket);
    } else {
      SOCKET_REFS.set(socket, count - 1);
    }
  }
}
```
this ref() and unref() are called once each 
the part of the stack trace here is 
```
ref (index.ts:833)
(anonymous) (index.ts:423)
C (sdk-3.0.0.js:8)
V (sdk-3.0.0.js:8)
execHttp2 (index.ts:408)
onConnect (index.ts:744)
onceWrapper (events.js:421)
emit (events.js:315)
EventEmitter.emit (domain.js:483)
onConnectSecure (_tls_wrap.js:1529)
emit (events.js:315)
EventEmitter.emit (domain.js:483)
TLSSocket._finishInit (_tls_wrap.js:936)
ssl.onhandshakedone (_tls_wrap.js:710)
```

**Comment by blakeembrey** [comment]
OK, I've been banging my head because I couldn't understand how it's possible for you to even get a `SocketSet`. The initial error report isn't very clear. I think this is the key issue:

> if i change the finally()'s to .then() (workaround)
> i now get a : socket.ref is not a function

So, to clarify, the issue was that `finally` didn't exist as a method? What was the error? Changing it from `.finally` to `.then` would affect the functionality and cause issues like you see, they have much different behaviors. `then` would have returned the value, `finally` discards the value (just like doing `finally` in normal code). Did you edit anything else? Can we tell me the error with a clean install?

**Comment by blakeembrey** [comment]
> but finally() must also be some ts to js compiler level dependency

Finally is just normal JavaScript FYI, and normal JavaScript I wrote: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally#browser_compatibility. It's been supported since node since v10, so it shouldn't be missing in electron.

**Comment by blakeembrey** [comment]
Are you using webpack or babel or something else? The only thing I could find online is maybe https://stackoverflow.com/questions/66491345/promise-finally-is-not-defined-in-electron-with-babel, could you be using non-native promises by accident? In any case, I've refactored this code on master to avoid using `.finally` and will release it now.

**Comment by blakeembrey** [comment]
Released as [1.1.3](https://github.com/serviejs/popsicle-transport-http/releases/tag/v1.1.3).

**Comment by sdetweil** [comment]
was not using webpack or babel.

the 1.1.3 version resolves the issue
